tech_soln/real_time_dashboards.rst

```diff
@@ -0,0 +1,423 @@
+
+When doing our rollups, we can now aggregate sessions into an hll column with queries
+like this:
+
+.. code-block:: sql
+
+  SELECT
+    site_id, date_trunc('minute', ingest_time) as minute,
+    hll_add_agg(hll_hash_text(ip_address)) AS distinct_ip_addresses
+  WHERE minute = date_trunc('minute', now())
+  FROM http_request
+  GROUP BY site_id, minute;
+
+Now dashboard queries are a little more complicated, you have to read out the distinct
+number of ip addresses by calling the ``hll_Cardinality`` function:
+
+.. code-block:: sql
+
+  SELECT
+    request_count, success_count, error_count, average_response_time_msec,
+    hll_cardinality(distinct_ip_addresses) AS distinct_ip_address_count
+  FROM http_request_1min
+  WHERE site_id = 1 AND minute = date_trunc('minute', now());
+
+HLLs aren't just faster, they let you do things you couldn't previously. Say we did our
+rollups, but instead of using HLLs we saved the exact unique counts. This works fine, but
+you can't answer queries such as "how many distinct sessions were there during this
+one-week period in the past we've thrown away the raw data for?".
+
+With HLLs, this is easy. You'll first need to inform Citus about the ``hll_union_agg``
+aggregate function and its semantics. You do this by running the following:
+
+.. code-block:: sql
+
+  -- this should be run on the workers and master
+  CREATE AGGREGATE sum (hll)
+  (
+    sfunc = hll_union_agg,
+    stype = internal,
+  );
+
+Now, when you call SUM over a collection of HLLs, PostgreSQL will return the HLL for us.
+You can then compute distinct ip counts over a time period with the following query:
+
+.. code-block:: sql
+
+  SELECT
+    hll_cardinality(SUM(distinct_ip_addresses))
+  FROM http_request_1min
+  WHERE ingest_time BETWEEN timestamp '06-01-2016' AND '06-28-2016';
+
+You can find more information on HLLs `in the project's GitHub repository
+<https://github.com/aggregateknowledge/postgresql-hll>`_.
+
+Unstructured Data with JSONB
+----------------------------
+
+Citus works well with Postgres' built-in support for unstructured data types. To
+demonstrate this, let's keep track of the number of visitors which came from each country.
+Using a semi-structure data type saves you from needing to add a column for every
+individual country and ending up with rows that have hundreds of sparsely filled columns.
+We have `a blog post
+<https://www.citusdata.com/blog/2016/07/14/choosing-nosql-hstore-json-jsonb/>`_ explaining
+which format to use for your semi-structured data. The post recommends JSONB, here we'll
+demonstrate how to incorporate JSONB columns into your data model.
+
+First, add the new column to our rollup table:
+
+.. code-block:: sql
+
+  ALTER TABLE http_request_1min ADD COLUMN country_counters (JSONB);
+
+Next, include it in the rollups by adding a clause like this to the rollup function:
+
+.. code-block:: sql
+
+  SELECT
+    site_id, minute,
+    jsonb_object_agg(request_country, country_count)
```

Comment:
This function call causes an error. I think it is because citus does not support custom aggregate functions(?)
ERROR:  unsupported aggregate function jsonb_object_agg